### PR TITLE
refactor(runtime/permissions): rewrite permission handling

### PR DESCRIPTION
**Pull request by crowlKats**

### CODE REVIEW

File: runtime/permissions.rs
Comment (nayeemrmn) [inline runtime/permissions.rs:71]:
```suggestion
  pub name: &'static str,
  #[serde(skip)]
  pub description: &'static str,
```

File: runtime/permissions.rs
Comment (nayeemrmn) [inline runtime/permissions.rs:454]:
```suggestion
  pub name: &'static str,
  pub description: &'static str,
```

**Review by nayeemrmn — COMMENTED** [review]
How does the `name`, `description` mechanism work for the [de]serialising in `worker_host.rs`? I don't think anything should be `#[serde(skip)]`ed here.

**Comment by crowlKats** [comment]
> How does the `name`, `description` mechanism work for the [de]serialising in `worker_host.rs`? I don't think anything should be `#[serde(skip)]`ed here.

well it creates a new instance of `Permissions` and merges it with the existing one. `name` & `description` are for handling of prompts, logs & errors. so these shouldn't be settable by the user.

File: runtime/permissions.rs
Comment (nayeemrmn) [inline runtime/permissions.rs:72]: What's the reason for this rename? This should be rigorously named `global_state`.
```suggestion
  pub global_state: PermissionState,
```
It doesn't make sense for `unary_permission.state` to be `PermissionState::Prompt` while `unary_permission.query(foo)` can return `PermissionState::Granted`.

Comment (crowlKats) [reply runtime/permissions.rs]: my bad, that was a mishap on my side

File: cli/tests/059_fs_relative_path_perm.ts.out
Comment (nayeemrmn) [inline cli/tests/059_fs_relative_path_perm.ts.out:1]: The message should contain the flag suggestion. The wording is a bit off for the global case as well. I sense that the capitalisation of the permission name is a problem? How about just:
```suggestion
[WILDCARD]error: Uncaught PermissionDenied: Requires read access to "non-existent", run again with --allow-read.
```

Comment (bartlomieju) [reply cli/tests/059_fs_relative_path_perm.ts.out:1]: I agree with Nayeem, the message should stay as is and ideally we could conditionally skip the part with flag to address #9002

Comment (crowlKats) [reply cli/tests/059_fs_relative_path_perm.ts.out]: @bartlomieju is it better now?
`Access to read "non-existent" required, run again with the --allow-read flag`

File: runtime/permissions.rs
Comment (bartlomieju) [inline runtime/permissions.rs:311]: This type looks funky, why use reference to option and then reference to tuple?

File: runtime/permissions.rs
Comment (bartlomieju) [inline runtime/permissions.rs:284]: Nitpick: might make more sense to have named fields?

File: runtime/permissions.rs
Comment (bartlomieju) [inline runtime/permissions.rs:351]: Nitpick: it's quite hard to read so many nested conditionals, maybe you could reorganize the code to return early instead?

File: runtime/permissions.rs
Comment (bartlomieju) [inline runtime/permissions.rs:388]: Ditto

File: runtime/permissions.rs
Comment (bartlomieju) [inline runtime/permissions.rs:466]: I'm not sure why it's called `UnaryPermission` instead of `ListPermission`...

File: runtime/permissions.rs
Comment (bartlomieju) [inline runtime/permissions.rs:681]: Not sure if we want to add `Deno requests`, opinions @ry?

**Review by bartlomieju — COMMENTED** [review]
Great work @crowlKats! I have a few small nitpicks regarding the implementation. Before landing I will wait for another review.

File: runtime/permissions.rs
Comment (nayeemrmn) [reply runtime/permissions.rs:466]: > I'm not sure why it's called `UnaryPermission` instead of `ListPermission`...

Queries for that permission take one parameter.

File: runtime/permissions.rs
Comment (crowlKats) [reply runtime/permissions.rs:311]: This is how it currently already is implemented. Not sure how simple it would be to change

File: runtime/permissions.rs
Comment (crowlKats) [reply runtime/permissions.rs:284]: Made this tuple just to alignments to what is supplied as arguments in the other net calls, so tuple. I can change it though if you prefer

File: runtime/permissions.rs
Comment (crowlKats) [reply runtime/permissions.rs:681]: This is the current, just that currently everywhere the function is called "Deno requests " is put in the request string. This is just deduplication

File: runtime/permissions.rs
Comment (bartlomieju) [reply runtime/permissions.rs:284]: Sounds reasonable

File: cli/tests/059_fs_relative_path_perm.ts.out
Comment (nayeemrmn) [reply cli/tests/059_fs_relative_path_perm.ts.out]: IMO they are still less concise/uniform than the original messages.
- `Access to read "foo" required`
- `Access to write to "foo" required` 🤮
- `Access to network for "foo:80" required`
- `Access to run a subprocess required`

vs original:
- `read access to "foo" required`
- `write access to "foo" required`
- `network access to "foo:80" required`
- `access to run a subprocess required`

My suggestion:
- `Requires read access to "foo"`
- `Requires write access to "foo"`
- `Requires network access to "foo:80"`
- `Requires subprocess access`

Comment (bartlomieju) [reply cli/tests/059_fs_relative_path_perm.ts.out]: I agree with Nayeem

File: runtime/permissions.rs
Comment (bartlomieju) [reply runtime/permissions.rs:681]: I see, then LGTM

File: cli/tests/059_fs_relative_path_perm.ts.out
Comment (crowlKats) [reply cli/tests/059_fs_relative_path_perm.ts.out]: `Requires read access to "foo"` sounds a bit off to me, as there is no subject for `Requires`. I know `deno` is implied, but it still sounds weird

File: runtime/ops/worker_host.rs
Comment (ry) [inline runtime/ops/worker_host.rs:35]: we prefer one line per import

File: runtime/permissions.rs
Comment (ry) [inline runtime/permissions.rs:71]: Why serde skip? Is the Deserialize trait even used?

Comment (crowlKats) [reply runtime/permissions.rs:71]: yes, for worker permissions

**Comment by ry** [comment]
Just to be clear - there's no change in behavior beyond some reworking of error messages?

**Comment by crowlKats** [comment]
> Just to be clear - there's no change in behavior beyond some reworking of error messages?

indeed. still works the same, just used differently. as proof you can see that the tests havent changed

File: cli/tests/059_fs_relative_path_perm.ts.out
Comment (nayeemrmn) [reply cli/tests/059_fs_relative_path_perm.ts.out]: > `Requires read access to "foo"` sounds a bit off to me, as there is no subject for `Requires`. I know `deno` is implied, but it still sounds weird

The implied subject is the operation that failed. I think plenty of error messages are structured like that.

Comment (crowlKats) [reply cli/tests/059_fs_relative_path_perm.ts.out]: I forgot for a moment that error message in JS would be `PermissionDenied: Requires read access to "foo", run again with --allow-read.`, so it would actually work. will rework the messages.

Comment (crowlKats) [reply cli/tests/059_fs_relative_path_perm.ts.out]: Alright, error messages now are 
```
Requires net access to "google.com", run again with the --allow-net flag
Requires read access to "./test.ts", run again with the --allow-read flag
Requires write access to "./test.ts", run again with the --allow-write flag
Requires env access, run again with the --allow-env flag
Requires run access, run again with the --allow-run flag
Requires plugin access, run again with the --allow-plugin flag
```
and debug log messages are
```
Granted net access to "www.google.com"
Granted read access to "./test.ts"
Granted write access to "./test.ts"
Granted env access
Granted run access
Granted plugin access
```
will this do?

File: runtime/permissions.rs
Comment (crowlKats) [reply runtime/permissions.rs]: changed to `Option<&(T, Option<u16>)>`

**Review by ry — APPROVED** [review]
LGTM - nice clean up @crowlKats and sorry for the delay.

**Pull request merged**
